File: src/lowlevel/client/icefsoperators/icefsListXattr.cpp
/*
 * @Author: Tan90degrees tangentninetydegrees@gmail.com
 * @Date: 2023-03-30 04:19:29
 * @LastEditors: Tan90degrees tangentninetydegrees@gmail.com
 * @LastEditTime: 2023-04-04 15:53:34
 * @FilePath: /icefs/src/lowlevel/client/icefsoperators/icefsListXattr.cpp
 * @Description:
 *
 * Copyright (C) 2023 Tan90degrees <tangentninetydegrees@gmail.com>.
 */
#include <stdio.h>

#include <string>

#include "icefsClient.hpp"
#include "icefsClientUtils.hpp"

void IcefsClient::DoIcefsListXattr(fuse_req_t fuseReq, fuse_ino_t inode,
                                   size_t size) {
  IcefsListXattrReq req;
  IcefsListXattrRes res;
  grpc::ClientContext ctx;
  ICEFS_PR_FUNCTION;
  req.set_inode(inode);
  req.set_size(size);

  grpc::Status status = stub_->DoIcefsListXattr(&ctx, req, &res);
  if (status.ok() && !res.status()) {
    if (size) {
      int32_t ret = fuse_reply_buf(fuseReq, res.value().c_str(), res.size());
      if (ret != ICEFS_EOK) {
        fuse_reply_err(fuseReq, ret);
        std::cout << "ERR:" << __FUNCTION__ << "fuse_reply_buf" << ret
                  << std::endl;
      }
    } else {
      fuse_reply_xattr(fuseReq, res.size());
    }
  } else {
    fuse_reply_err(fuseReq, res.status() ? res.status() : EIO);
    ICEFS_PR_ERR_STATUS;
  }
}

File: src/lowlevel/server/icefsoperators/icefsListXattr.go
/*
 * @Author: Tan90degrees tangentninetydegrees@gmail.com
 * @Date: 2023-03-11 07:18:32
 * @LastEditors: Tan90degrees tangentninetydegrees@gmail.com
 * @LastEditTime: 2023-04-18 05:57:27
 * @FilePath: /icefs/src/lowlevel/server/icefsoperators/icefsListXattr.go
 * @Description:
 *
 * Copyright (C) 2023 Tan90degrees <tangentninetydegrees@gmail.com>.
 */
package icefsoperators

import (
	"context"
	"fmt"
	"icefs-server/icefserror"
	pb "icefs-server/icefsgrpc"
	"icefs-server/icefsthrift"
	"syscall"
)

func (s *IcefsServer) doIcefsListXattr(fakeInode uint64, reqSize uint64) (status int32, resSize int64, value []byte) {
	var procName string
	var size int
	var err error
	s.inodeCacheLock.RLock()
	inode := s.getIcefsInode(fakeInode)
	if inode == nil {
		s.inodeCacheLock.RUnlock()
		status = icefserror.ICEFS_BUG_ERR
		return
	}
	inode.inodeLock.RLock()
	s.inodeCacheLock.RUnlock()
	procName = fmt.Sprintf("/proc/self/fd/%v", inode.fd)
	inode.inodeLock.RUnlock()
	value = make([]byte, reqSize)
	size, err = syscall.Listxattr(procName, value)
	if err != nil {
		status = icefserror.IcefsStdErrno(err)
		return
	}
	status = icefserror.ICEFS_EOK
	resSize = int64(size)
	return
}

func (s *IcefsGRpcServer) DoIcefsListXattr(ctx context.Context, req *pb.IcefsListXattrReq) (*pb.IcefsListXattrRes, error) {
	var res pb.IcefsListXattrRes

	res.Status, res.Size, res.Value = s.server.doIcefsListXattr(req.Inode, req.Size)

	return &res, nil
}

func (s *IcefsThriftServer) DoIcefsListXattr(ctx context.Context, req *icefsthrift.IcefsListXattrReq) (*icefsthrift.IcefsListXattrRes, error) {
	var res icefsthrift.IcefsListXattrRes

	res.Status, res.Size, res.Value = s.server.doIcefsListXattr(uint64(req.Inode), uint64(req.Size))

	return &res, nil
}
